pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,790 @@
+				for _, s := range secretObjects {
+					secretObject, ok := s.(map[string]interface{})
+					if !ok {
+						continue
+					}
+
+					secretName, err := getStringFromObject(secretObject, "secretName")
+					if err != nil {
+						continue
+					}
+					err = deleteK8sSecret(ctx, secretName, namespace)
+					if err != nil {
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+		}
+
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, deleteSecretFn); err != nil {
+			log.Error(err, "max retries for deleting secret reached for pod: %s, ns: %s", podUID, namespace)
+			return err
+		}
+	}
+	return nil
+}
+
+// createOrUpdateK8sSecret creates or updates a K8s secret with data from mounted files
+// If a secret with the same name already exists in the namespace of the pod, it's updated.
+func createOrUpdateK8sSecret(ctx context.Context, name string, namespace string, datamap map[string][]byte, secretType corev1.SecretType) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Type: secretType,
+	}
+
+	err = c.Get(ctx, secretKey, secret)
+	if err != nil {
+		log.Error(err, "error from c.Get for secret: %s, ns: %s", name, namespace)
+		if errors.IsNotFound(err) {
+			secret.Data = datamap
+			if err := c.Create(ctx, secret); err != nil {
+				log.Error(err, "error while creating K8s secret: %s, ns: %s", name, namespace)
+				return err
+			}
+		} else {
+			log.Error(err, "error while retrieving K8s secret: %s, ns: %s", name, namespace)
+			return err
+		}
+	}
+	secret.Data = datamap
+	if err := c.Update(ctx, secret); err != nil {
+		log.Error(err, "error while updating K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	log.Infof("created k8s secret: %s, ns: %s", name, namespace)
+	return nil
+}
+
+// deleteK8sSecret deletes a secret by name
+func deleteK8sSecret(ctx context.Context, name string, namespace string) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		if errors.IsNotFound(err) {
+			log.Infof("k8s secret not found during delete. Skip. secret: %s, ns: %s", name, namespace)
+			return nil
+		}
+		log.Error(err, "error while getting K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	if err := c.Delete(ctx, secret); err != nil {
```

Comment:
On second thought, wouldn't `c.Delete` return `secret not found` error if the secret doesn't exist? So can we just do a delete call instead of get and then delete? If error is `secret not found`,  then we can just `return nil`.